src/NHibernate/Linq/LinqExtensionMethods.cs

```diff
@@ -2463,6 +2464,97 @@ public static IFutureValue<TSource> ToFutureValue<TSource>(this IQueryable<TSour
+		/// <param name="outerKeySelector">A dynamic function to extract the join key from each element of the first sequence.</param>
+		/// <param name="innerKeySelector">A dynamic function to extract the join key from each element of the second sequence.</param>
+		/// <param name="resultSelector">A dynamic function to create a result element from two matching elements.</param>
+		/// <returns>An <see cref="IQueryable"/> obtained by performing a left join on two sequences.</returns>
+		public static IQueryable<TResult> LeftJoin<TOuter, TInner, TKey, TResult>(
+			this IQueryable<TOuter> outer,
+			IQueryable<TInner> inner,
+			Expression<Func<TOuter, TKey>> outerKeySelector,
+			Expression<Func<TInner, TKey>> innerKeySelector,
+			Expression<Func<TOuter, TInner, TResult>> resultSelector)
+		{
+			outer = outer ?? throw new ArgumentNullException(nameof(outer));
+			inner = inner ?? throw new ArgumentNullException(nameof(inner));
+			outerKeySelector = outerKeySelector ?? throw new ArgumentNullException(nameof(outerKeySelector));
+			innerKeySelector = innerKeySelector ?? throw new ArgumentNullException(nameof(innerKeySelector));
+			resultSelector = resultSelector ?? throw new ArgumentNullException(nameof(resultSelector));
+
+			Expression<Func<TOuter, IEnumerable<TInner>, LeftJoinIntermediate<TOuter, TInner>>> groupJoinResultSelector =
+				(oneOuter, manyInners) => new LeftJoinIntermediate<TOuter, TInner>
+				{
+					OneOuter = oneOuter,
+					ManyInners = manyInners
+				};
+			var groupJoin = GroupJoinDefinition.MakeGenericMethod(
+				typeof(TOuter),
+				typeof(TInner),
+				typeof(TKey),
+				typeof(LeftJoinIntermediate<TOuter, TInner>));
+			var selectMany = SelectManyDefinition.MakeGenericMethod(
+				typeof(LeftJoinIntermediate<TOuter, TInner>),
+				typeof(TInner),
+				typeof(TResult));
+			var exprGroupJoin = Expression.Call(
+				groupJoin,
+				outer.Expression,
+				inner.Expression,
+				outerKeySelector,
+				innerKeySelector,
+				groupJoinResultSelector);
+			var selectManyCollectionSelector = (Expression<Func<LeftJoinIntermediate<TOuter, TInner>, IEnumerable<TInner>>>)
+				(t => t.ManyInners.DefaultIfEmpty());
+			var outerParameter = resultSelector.Parameters[0];
+			var paramNew = Expression.Parameter(typeof(LeftJoinIntermediate<TOuter, TInner>));
+			var outerProperty = Expression.Property(paramNew, nameof(LeftJoinIntermediate<TOuter, TInner>.OneOuter));
+			var selectManyResultSelector = Expression.Lambda(
+				new Replacer(outerParameter, outerProperty).Visit(resultSelector.Body),
+				paramNew,
+				resultSelector.Parameters[1]);
+
+			return outer.Provider.CreateQuery<TResult>(
+				Expression.Call(selectMany, exprGroupJoin, selectManyCollectionSelector, selectManyResultSelector));
+		}
+
+		private class LeftJoinIntermediate<TOuter, TInner>
+		{
+			public TOuter OneOuter { get; set; }
+			public IEnumerable<TInner> ManyInners { get; set; }
+		}
+
+		private class Replacer : ExpressionVisitor
```

Comment:
Re-linq already has a class which does this. `ReplacingSomethingVisitor`.

Reply:
I didn't spot that before. Replaced with Relinq `ReplacingExpressionVisitor`.